### ISSUE TITLE
Format with another type or relation by FK

### ISSUE THREAD

**Issue by knopa** [thepirat000/Audit.NET#526]
I have two question regarding EntityFramework for .net core 6  version
1. Is it possible to set format for Changes Old, New Value like in EntityFramework-Plus

`.Format<OrderItem>(x => x.Price,
                                      x => x.ToString("$#.00"));`
As example for enum types which stored as int but you want to show them as string.
2. Is possible to grab more info for related object into to changes
Like I have 
User -> UserProfile
UserProfile -> MentorId -> User
User -> UserGroup
If I have changed MentorId - be able to Format value to use Mentor.Name or fetch it

**Comment by knopa** [comment]
I gues with first found a solution but I think itsn't not useful - I think better to do more flexible current format

`_auditContext.EntitySettings = new()
                {
                    {
                        typeof(UserProfile), 
                        new EfEntitySettings()
                        {
                            FormatProperties = new Dictionary<string, Func<object, object>>()
                            {
                                {nameof(UserProfile.EmployeeType), val => ((EmployeeType)val).ToString()}
                            }
                        }
                    }
                };`

**Comment by thepirat000** [comment]
For the format, there is a Format method actually:

https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#settings-high-level-interceptor

```c#
Audit.EntityFramework.Configuration.Setup()
    .ForContext<MyContext>(config => config
        .ForEntity<OrderItem>(_ => _
            .Format(x => x.Price, x => x.ToString("$#.00"))));
```

**Comment by knopa** [comment]
Well it doesn't work as expected because it use same Type as prop not object one

`Format<TProp>(Expression<Func<TEntity, TProp>> property, Func<TProp, TProp> format);`

**Comment by knopa** [comment]
I have created PR with that change if you don't mind https://github.com/thepirat000/Audit.NET/pull/527

**Comment by thepirat000** [comment]
Oh, you're right

Of course I don't mind, thanks a lot!

Will include the changes on the next release

**Comment by knopa** [comment]
What is ETA for that?

Any suggestion on second option?

**Comment by thepirat000** [comment]
Maybe today.

Could you elaborate a little bit more on your second question? I think I don't fully understand

**Comment by thepirat000** [comment]
Which data provider are you using to store the audit events?

**Comment by knopa** [comment]
Audit.EntityFramework

**Comment by thepirat000** [comment]
Are you storing the audits on tables in the same database you are auditing?
Could you share your data provider configuration?
`Audit.Core.Configuration.Setup().UseEntityFramework(...)`

**Comment by knopa** [comment]
It looks like that

``` csharp
Audit.Core.Configuration
            .AddCustomAction(ActionType.OnScopeCreated, scope =>
            {
                var id = Guid.NewGuid();
                scope.SetCustomField("AuditScopeId", id);
            });
```

``` csharp
Audit.Core.Configuration.Setup()
    .UseEntityFramework(_ =>_
        .AuditTypeMapper(_ => typeof(AuditLog))
        .AuditEntityAction<AuditLog>(Handle)
        .IgnoreMatchedProperties());


private static void Handle(AuditEvent ev, EventEntry entry, AuditLog target)
    {
        if (ev.CustomFields.TryGetValue("AuditScopeId", out object value))
        {
            target.TransactionId = (Guid)value;
        }

        AuditConfig config = new();
        if (ev.CustomFields.TryGetValue(nameof(AuditConfig), out object configValue))
        {
            config = (AuditConfig)configValue;
        }

        target.Event = config.Event;
        target.RootEntityId = config.RootEntityId;
        target.RootEntityType = config.RootEntityType;
        target.PerformedById = config.ResponsibleId;

        target.EntityType = entry.EntityType.Name;
        target.AuditDate = DateTime.UtcNow;
        target.ActionType = entry.Action;
        target.EntityId = entry.PrimaryKey.First().Value.ToString();

        target.Changes = new();
        if ((target.ActionType == Core.Models.Enums.ActionType.Add
         || target.ActionType == Core.Models.Enums.ActionType.Delete)
            && entry.ColumnValues != null)
        {
            target.Changes.AddRange(entry.ColumnValues.Select(c => new PropertyChange()
            {
                Name = c.Key,
                New = c.Value
            }));
        }

        if (entry.Changes != null)
        {
            target.Changes.AddRange(entry.Changes.Select(c => new PropertyChange()
            {
                Name = c.ColumnName,
                New = c.NewValue,
                Old = c.OriginalValue
            }));
        }
    }

**Comment by thepirat000** [comment]
If you really need to do it with a method like `Format`, which replaces the values on Changes and ColumnValues collection, then I think it could be possible to provide another overload that takes an [EntityEntry](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.changetracking.entityentry?view=efcore-6.0) instead of (of apart from) the current value. 

So you could do:

```c#
  .ForEntity<User>(cfg => cfg
    .Format(u => u.UserName, 
      entry => ((MyContext)entry.Context).Teachers.First(t => t.Id == ((User)entry.Entity).Id).TeacherName
```

The sad part is that you'll probably need to do some casting.

Do you think something like that will work for you?

**Comment by knopa** [comment]
Looks not bad. You be able to grab value from entry relation or from context if it not included

**Comment by thepirat000** [comment]
This was implemented in version 19.4.0, please upgrade your references and re-test.

- Relaxed `Format()` method to allow returning any type.
- New overload for `Override()` method taking an EntityEntry

**Comment by knopa** [comment]
Great
